### ISSUE TITLE
Workbench auto save errors when spatial plugin installed

### ISSUE THREAD

**Issue by jonescc** [talend-spatial/talend-spatial#77]
After installing the spatial plugin in Talend Open Studio 7.1.1, I sometimes get LinkageErrors every time a workbench auto save is performed in some installations (depends on whether the org.eclipse.emf.common.util is decided to sourced from org.talend.libraries.sdi instead of org.eclipse.emf.common when TOS is loaded the first time).

**Comment by jonescc** [comment]
This happens for me every time when installed using vagrant with Ubuntu 16.04 base box.  It doesn't happen for me in virtualbox with Ubuntu 16.04 desktop installed or in docker with 18.04 installed (well perhaps it did, I vaguely recall getting auto save problems but it went away and I didn't worry about it).

Regardless, its an issue to export packages like this when they should be sourced from the core plugins (https://developer.atlassian.com/server/framework/atlassian-sdk/linkageerror/)

**Comment by jonescc** [comment]
Screen shot of error in TOS 7.1.1:

![image](https://user-images.githubusercontent.com/1860215/59577772-5b5b8080-9108-11e9-9792-a43bfd0caf85.png)

Extract from log:

```
!ENTRY org.eclipse.core.jobs 4 2 2019-06-17 13:59:58.223
!MESSAGE An internal error occurred during: "Workbench Auto-Save Background Job".
!STACK 0
java.lang.LinkageError: org/eclipse/emf/common/util/EList
        at org.eclipse.ui.internal.Workbench$18.run(Workbench.java:1301)
        at org.eclipse.core.internal.jobs.Worker.run(Worker.java:54)
```

Class loader debug output when error occurs:

```
ModuleClassLoader[org.eclipse.ui.workbench_3.106.1.v20140827-1737].loadClass(org.eclipse.ui.XMLMemento$DOMWriter)
BundleLoader[org.eclipse.ui.workbench_3.106.1.v20140827-1737].findClassInternal(org.eclipse.ui.XMLMemento$DOMWriter)
BundleLoader[org.eclipse.ui.workbench_3.106.1.v20140827-1737].findLocalClass(org.eclipse.ui.XMLMemento$DOMWriter)
ModuleClassLoader[org.eclipse.ui.workbench_3.106.1.v20140827-1737 - /var/lib/talend7/plugins/org.eclipse.ui.workbench_3.106.1.v20140827-1737.jar].findClassImpl(org.eclipse.ui.XMLMemento$DOMWriter)
  read 3615 bytes from /var/lib/talend7/plugins/org.eclipse.ui.workbench_3.106.1.v20140827-1737.jar!/org/eclipse/ui/XMLMemento$DOMWriter.class
  defining class org.eclipse.ui.XMLMemento$DOMWriter
ModuleClassLoader[org.eclipse.ui.workbench_3.106.1.v20140827-1737].loadClass(java.io.PrintWriter)
BundleLoader[org.eclipse.ui.workbench_3.106.1.v20140827-1737] found local class org.eclipse.ui.XMLMemento$DOMWriter
...
ModuleClassLoader[org.eclipse.ui.workbench_3.106.1.v20140827-1737].loadClass(org.eclipse.emf.common.util.EList)
BundleLoader[org.eclipse.ui.workbench_3.106.1.v20140827-1737].findClassInternal(org.eclipse.emf.common.util.EList)
BundleLoader[org.eclipse.ui.workbench_3.106.1.v20140827-1737] loading from import package: org.eclipse.emf.common.util -> [org.talend.libraries.sdi_7.1.1]
BundleLoader[org.talend.libraries.sdi_7.1.1].findLocalClass(org.eclipse.emf.common.util.EList)
ModuleClassLoader[org.talend.libraries.sdi_7.1.1 - /var/lib/talend7/plugins/org.talend.libraries.sdi-7.1.1].findClassImpl(org.eclipse.emf.common.util.EList)
ModuleClassLoader[org.talend.libraries.sdi_7.1.1 - /var/lib/talend7/plugins/org.talend.libraries.sdi-7.1.1/lib/jsr-275-1.0-beta-2.jar].findClassImpl(org.eclipse.emf.common.util.EList)
ModuleClassLoader[org.talend.libraries.sdi_7.1.1 - /var/lib/talend7/plugins/org.talend.libraries.sdi-7.1.1/lib/gt-opengis-8.5.jar].findClassImpl(org.eclipse.emf.common.util.EList)
ModuleClassLoader[org.talend.libraries.sdi_7.1.1 - /var/lib/talend7/plugins/org.talend.libraries.sdi-7.1.1/lib/gt-api-8.5.jar].findClassImpl(org.eclipse.emf.common.util.EList)
ModuleClassLoader[org.talend.libraries.sdi_7.1.1 - /var/lib/talend7/plugins/org.talend.libraries.sdi-7.1.1/lib/gt-main-8.5.jar].findClassImpl(org.eclipse.emf.common.util.EList)
ModuleClassLoader[org.talend.libraries.sdi_7.1.1 - /var/lib/talend7/plugins/org.talend.libraries.sdi-7.1.1/lib/gt-metadata-8.5.jar].findClassImpl(org.eclipse.emf.common.util.EList)
ModuleClassLoader[org.talend.libraries.sdi_7.1.1 - /var/lib/talend7/plugins/org.talend.libraries.sdi-7.1.1/lib/gt-referencing-8.5.jar].findClassImpl(org.eclipse.emf.common.util.EList)
ModuleClassLoader[org.talend.libraries.sdi_7.1.1 - /var/lib/talend7/plugins/org.talend.libraries.sdi-7.1.1/lib/gt-shapefile-8.5.jar].findClassImpl(org.eclipse.emf.common.util.EList)
ModuleClassLoader[org.talend.libraries.sdi_7.1.1 - /var/lib/talend7/plugins/org.talend.libraries.sdi-7.1.1/lib/gt-wfs-8.5.jar].findClassImpl(org.eclipse.emf.common.util.EList)
ModuleClassLoader[org.talend.libraries.sdi_7.1.1 - /var/lib/talend7/plugins/org.talend.libraries.sdi-7.1.1/lib/hsqldb-1.8.0.7.jar].findClassImpl(org.eclipse.emf.common.util.EList)
ModuleClassLoader[org.talend.libraries.sdi_7.1.1 - /var/lib/talend7/plugins/org.talend.libraries.sdi-7.1.1/lib/commons-httpclient-3.1.jar].findClassImpl(org.eclipse.emf.common.util.EList)
ModuleClassLoader[org.talend.libraries.sdi_7.1.1 - /var/lib/talend7/plugins/org.talend.libraries.sdi-7.1.1/lib/dom4j-1.6.1.jar].findClassImpl(org.eclipse.emf.common.util.EList)
ModuleClassLoader[org.talend.libraries.sdi_7.1.1 - /var/lib/talend7/plugins/org.talend.libraries.sdi-7.1.1/lib/jdom-1.0.jar].findClassImpl(org.eclipse.emf.common.util.EList)
ModuleClassLoader[org.talend.libraries.sdi_7.1.1 - /var/lib/talend7/plugins/org.talend.libraries.sdi-7.1.1/lib/jts-1.12.jar].findClassImpl(org.eclipse.emf.common.util.EList)
ModuleClassLoader[org.talend.libraries.sdi_7.1.1 - /var/lib/talend7/configuration/.m2/repository/org/talend/libraries/org.talend.sdi/6.0.0/org.talend.sdi-6.0.0.jar].findClassImpl(org.eclipse.emf.common.util.EList)
ModuleClassLoader[org.talend.libraries.sdi_7.1.1 - /var/lib/talend7/plugins/org.talend.libraries.sdi-7.1.1/lib/common-2.6.0.jar].findClassImpl(org.eclipse.emf.common.util.EList)
  read 334 bytes from /var/lib/talend7/plugins/org.talend.libraries.sdi-7.1.1/lib/common-2.6.0.jar!/org/eclipse/emf/common/util/EList.class
  defining class org.eclipse.emf.common.util.EList
ModuleClassLoader[org.talend.libraries.sdi_7.1.1].loadClass(java.util.List)
BundleLoader[org.talend.libraries.sdi_7.1.1] found local class org.eclipse.emf.common.util.EList
```

Class loader debug output when error doesn't occur:

```
ModuleClassLoader[org.eclipse.ui.workbench_3.106.1.v20140827-1737].loadClass(org.eclipse.ui.XMLMemento$DOMWriter)
BundleLoader[org.eclipse.ui.workbench_3.106.1.v20140827-1737].findClassInternal(org.eclipse.ui.XMLMemento$DOMWriter)
BundleLoader[org.eclipse.ui.workbench_3.106.1.v20140827-1737].findLocalClass(org.eclipse.ui.XMLMemento$DOMWriter)
ModuleClassLoader[org.eclipse.ui.workbench_3.106.1.v20140827-1737 - /home/craigj/software/talend7/plugins/org.eclipse.ui.workbench_3.106.1.v20140827-1737.jar].findClassImpl(org.eclipse.ui.XMLMemento$DOMWriter)
  read 3615 bytes from /home/craigj/software/talend7/plugins/org.eclipse.ui.workbench_3.106.1.v20140827-1737.jar!/org/eclipse/ui/XMLMemento$DOMWriter.class
  defining class org.eclipse.ui.XMLMemento$DOMWriter
ModuleClassLoader[org.eclipse.ui.workbench_3.106.1.v20140827-1737].loadClass(java.io.PrintWriter)
BundleLoader[org.eclipse.ui.workbench_3.106.1.v20140827-1737] found local class org.eclipse.ui.XMLMemento$DOMWriter
...
ModuleClassLoader[org.eclipse.ui.workbench_3.106.1.v20140827-1737].loadClass(org.eclipse.emf.common.util.EList)
BundleLoader[org.eclipse.ui.workbench_3.106.1.v20140827-1737].findClassInternal(org.eclipse.emf.common.util.EList)
BundleLoader[org.eclipse.ui.workbench_3.106.1.v20140827-1737] loading from import package: org.eclipse.emf.common.util -> [org.eclipse.emf.common_2.10.1.v20140901-1043]
BundleLoader[org.eclipse.emf.common_2.10.1.v20140901-1043].findLocalClass(org.eclipse.emf.common.util.EList)
BundleLoader[org.eclipse.emf.common_2.10.1.v20140901-1043] found local class org.eclipse.emf.common.util.EList
```

**Comment by jonescc** [comment]
So debugging this further, on the installation where this problem occurs, it looks like both bundles are selected as candidates for sourcing org.eclipse.emf.common.util, but the sdi version is listed first which causes the issue.

```
RESOLVER: Finding capabilities for requirement
	osgi.wiring.package; filter:="(osgi.wiring.package=org.eclipse.emf.common.util)"
		of resource
			osgi.identity; osgi.identity="org.eclipse.ui.workbench"; type="osgi.bundle"; version:Version="3.106.1.v20140827-1737"; singleton:="true"
RESOLVER: Capabilities being returned to the resolver
	[1] osgi.wiring.package; bundle-symbolic-name="org.talend.libraries.sdi"; bundle-version:Version="7.1.1"; version:Version="0.0.0"; osgi.wiring.package="org.eclipse.emf.common.util"
		of resource
			osgi.identity; osgi.identity="org.talend.libraries.sdi"; type="osgi.bundle"; version:Version="7.1.1"; singleton:="true"
	[2] osgi.wiring.package; bundle-symbolic-name="org.eclipse.emf.common"; bundle-version:Version="2.10.1.v20140901-1043"; version:Version="0.0.0"; osgi.wiring.package="org.eclipse.emf.common.util"
		of resource
			osgi.identity; osgi.identity="org.eclipse.emf.common"; type="osgi.bundle"; version:Version="2.10.1.v20140901-1043"; singleton:="true"
```

The sdi version is listed first because it was loaded first when starting up the framework.

Debugging into the eclipse code, the order in which the plugins are loaded is determined by the order that the plugins directories/jar files are returned when listing the plugin directory (https://github.com/eclipse/eclipse.platform/blob/master/update/org.eclipse.update.configurator/src/org/eclipse/update/internal/configurator/SiteEntry.java#L317).  The plugin directory is listed using java.io.File.listFiles().  According to the documentation for this method (https://docs.oracle.com/javase/7/docs/api/java/io/File.html#listFiles()), the order that the files in the directory are returned is not guaranteed to be in any specific order.

So currently, the occurrence of this issue entirely depends on the order that the plugins are returned by listFiles which is not guaranteed to be in any specific order (its likely to be dependent on file system on which the files are stored and what actions have been performed on it already).

**Comment by fxprunayre** [comment]
Hi Craig, not sure I can help much here. So your commit https://github.com/jonescc/talend-spatial/commit/1cc35b121bfa7497a587583edca029998ea230a0 does not fix the issue, right ?
Why is there an error if the version of org.eclipse.emf.common is the same ?

**Comment by jonescc** [comment]
Yes my pull request (jonescc@1cc35b1) fixes the issue.  I was just clarifying why it works in one installation and not another above.   There's no need to export the package in the org.talend.libraries.sdi bundle as it's already available in the org.eclipse.emf.common bundle.  I switched it to an import so that it would still be available to anything that needs it in org.talend.libraries.sdi.

As I understand it, the problem is that it is loaded by a different class loader even though it its the same version of the class.   I'll dig up some references.   I found some good explanations a couple of days ago but didn't record them.

**Comment by jonescc** [comment]
https://medium.com/@technospace/java-vs-osgi-class-loading-17a1ad4cc2a5#9e81

**Comment by jonescc** [comment]
The key here is to ensure there is only one bundle which loads the class.  Before my change there were two possible bundles which could load the class and the error occurred when it was loaded from both.  Its not possible to get the linkageError after my change as only the org.eclipse.emf.common bundle will ever load the class.

**Comment by jonescc** [comment]
Another source of possible problems is that the versions of the jars were different.  In 7.1.1 the bundle is 
 org.eclipse.emf.common_2.10.1.v20140901-1043.jar versus the common-2.6.0.jar include in org.talend.libraries.sdi.

**Comment by jonescc** [comment]
I'm only new at this so I may be missing something, but the fix did work for me.

**Comment by jonescc** [comment]
To show that the problem is not the different versions of the jar that are being included, I've modified the 7.1.1 code to use the same version of org.eclipse.emf.common in the spatial plugin as installed with TOS 7.1.1 (2.10.1).  You can see the changes I made at https://github.com/talend-spatial/talend-spatial/compare/7.1.1...jonescc:align-version.  The changes were made by:

- replacing all common-2.6.0.jars in the source code with the TOS 7.1.1 org.eclipse.emf.common_2.10.1.v20140901-1043.jar renamed to common-2.10.1.jar and updating all common-2.6.0.jar references to refer to it instead (as appears to have been done with the other core plugin libraries)
- copying org.talend.libraries.sdi.jar built by maven clean package to main/plugins/org.talend.sdi.designer.components/components/sGeoInput/org.talend.sdi.jar and main/plugins/org.talend.sdi.repository.ui.actions.metadata/lib/org.talend.sdi.jar to ensure these jars also referred only to common-2.10.1.jar (updating these jars isn't done automatically when building the project)

ecore-2.6.1.jar and xsd-2.6.0.jar match the versions of the core jars so I didn't change these.

I then rebuilt TOS-Spatial-7.1.1.zip using ./package.sh 7.1.1. The result can be downloaded from https://s3-ap-southeast-2.amazonaws.com/imos-binary/static/talend/TOS-Spatial-7.1.1-patch-3.zip

I discovered I can use docker to control the order of bundle loading as later layers appear to be listed after earlier layers in directory listings so the problem is reproducible using docker.  This was working in my docker containers because I unzip TOS-Spatial-7.1.1.zip in a separate layer after unzipping the core TOS package.  I switched the order of this unzipping (https://github.com/jonescc/tos-7.1.1-docker/tree/tos-spatial-issue-demo) and the problem became even worse than just the auto-save error above (and matched what was happening on my desktop which I thought was caused by installing the cinnamon desktop instead of unity, but appears to be caused by talend-spatial):

![image](https://user-images.githubusercontent.com/1860215/59973459-961b5800-95e3-11e9-9e56-eeb3cb352282.png)

When I install the version of TOS-Spatial-7.1.1.zip built from my fix at https://github.com/talend-spatial/talend-spatial/pull/78 instead.  TOS starts up normally and there are no errors saving workbench settings.  I need to make a few changes to the pull request after going through the talend-spatial code more thoroughly for this testing, but sourcing the org.eclipse.emf.common package only from the core bundle fixes the issue.

**Comment by jonescc** [comment]
Also refer to the section on linkageerrors in https://chamibuddhika.wordpress.com/2012/05/15/learn-by-errors-java-osgi/
